# Clustering de perfiles de OkCupid
Este notebook muestra el proceso de análisis, preprocesamiento y entrenamiento de modelos para agrupar perfiles similares de la web de citas OkCupid.

## 1. Setup

Librerías importadas en el proyecto.

In [1]:
import numpy as np
import pandas as pd

import os

from IPython.display import IFrame

try:
    from geopy.geocoders import Nominatim
except ModuleNotFoundError:
    %pip install geopy
    from geopy.geocoders import Nominatim

%matplotlib inline
from matplotlib import pyplot as plt
import seaborn as sns

Variables globales declaradas en el proyecto.

In [2]:
RANDOM_SEED = 1337
np.random.seed(RANDOM_SEED)

#!git clone https://github.com/dietrujillo/ml_lab.git ./ml_lab
# os.path.join(os.curdir, "ml_lab", "lab3", "data")  # Path to csv data
DATA_PATH = os.path.join(os.pardir, "data")
# os.path.join(os.curdir, "ml_lab", "lab3", "info")  # Path to problem info
INFO_PATH = os.path.join(os.pardir, "info")
# os.path.join(os.curdir, "ml_lab", "lab3", "results")  # Path to results files
RESULTS_PATH = os.path.join(os.pardir, "results")

Enunciado del problema.

In [3]:
# Jupyter Notebook only
IFrame(os.path.join(INFO_PATH, "lab3.pdf"), width=1080, height=920)

## 2. Análisis exploratorio de los datos

In [4]:
pd.set_option("display.max_columns", 50)
df = pd.read_csv(os.path.join(DATA_PATH, "okcupid_profiles.csv"))
display(df.iloc[:, :10])
display(df.iloc[:, 10:20])
display(df.iloc[:, 20:31])

age     status sex orientation       body_type               diet  \
0       22     single   m    straight  a little extra  strictly anything   
1       35     single   m    straight         average       mostly other   
2       38  available   m    straight            thin           anything   
3       23     single   m    straight            thin         vegetarian   
4       29     single   m    straight        athletic                NaN   
...    ...        ...  ..         ...             ...                ...   
59941   59     single   f    straight             NaN                NaN   
59942   24     single   m    straight             fit    mostly anything   
59943   42     single   m    straight         average    mostly anything   
59944   27     single   m    straight        athletic    mostly anything   
59945   39     single   m         gay         average                NaN   

           drinks      drugs                          education  \
0        socially      never      working on college/university   
1           often  sometimes              working on space camp   
2        socially        NaN     graduated from masters program   
3        socially        NaN      working on college/university   
4        socially      never  graduated from college/university   
...           ...        ...                                ...   
59941    socially      never  graduated from college/university   
59942       often  sometimes      working on college/university   
59943  not at all      never     graduated from masters program   
59944    socially      often      working on college/university   
59945    socially        NaN     graduated from masters program   

                 ethnicity  
0             asian, white  
1                    white  
2                      NaN  
3                    white  
4      asian, black, other  
...                    ...  
59941                  NaN  
59942         white, other  
59943                asian  
59944         asian, black  
59945                white  

[59946 rows x 10 columns]

height  income                           job       last_online  \
0        75.0      -1                transportation  2012-06-28-20-30   
1        70.0   80000          hospitality / travel  2012-06-29-21-41   
2        68.0      -1                           NaN  2012-06-27-09-10   
3        71.0   20000                       student  2012-06-28-14-22   
4        66.0      -1   artistic / musical / writer  2012-06-27-21-26   
...       ...     ...                           ...               ...   
59941    62.0      -1   sales / marketing / biz dev  2012-06-12-21-47   
59942    72.0      -1         entertainment / media  2012-06-29-11-01   
59943    71.0  100000  construction / craftsmanship  2012-06-27-23-37   
59944    73.0      -1             medicine / health  2012-06-23-13-01   
59945    68.0      -1             medicine / health  2012-06-29-00-42   

                              location  \
0      south san francisco, california   
1                  oakland, california   
2            san francisco, california   
3                 berkeley, california   
4            san francisco, california   
...                                ...   
59941              oakland, california   
59942        san francisco, california   
59943  south san francisco, california   
59944        san francisco, california   
59945        san francisco, california   

                                    offspring                       pets  \
0      doesn't have kids, but might want them  likes dogs and likes cats   
1      doesn't have kids, but might want them  likes dogs and likes cats   
2                                         NaN                   has cats   
3                           doesn't want kids                 likes cats   
4                                         NaN  likes dogs and likes cats   
...                                       ...                        ...   
59941                                has kids                   has dogs   
59942                       doesn't have kids  likes dogs and likes cats   
59943                       doesn't have kids                        NaN   
59944       doesn't have kids, but wants them  likes dogs and likes cats   
59945                                     NaN  likes dogs and likes cats   

                                        religion  \
0          agnosticism and very serious about it   
1       agnosticism but not too serious about it   
2                                            NaN   
3                                            NaN   
4                                            NaN   
...                                          ...   
59941   catholicism but not too serious about it   
59942                                agnosticism   
59943  christianity but not too serious about it   
59944   agnosticism but not too serious about it   
59945          catholicism and laughing about it   

                                           sign          smokes  
0                                        gemini       sometimes  
1                                        cancer              no  
2            pisces but it doesn&rsquo;t matter              no  
3                                        pisces              no  
4                                      aquarius              no  
...                                         ...             ...  
59941  cancer and it&rsquo;s fun to think about              no  
59942           leo but it doesn&rsquo;t matter              no  
59943   sagittarius but it doesn&rsquo;t matter              no  
59944     leo and it&rsquo;s fun to think about  trying to quit  
59945  gemini and it&rsquo;s fun to think about       sometimes  

[59946 rows x 10 columns]

speaks  \
0                                                english   
1      english (fluently), spanish (poorly), french (...   
2                                   english, french, c++   
3                               english, german (poorly)   
4                                                english   
...                                                  ...   
59941                                            english   
59942                                 english (fluently)   
59943                                 english (fluently)   
59944  english (fluently), spanish (poorly), chinese ...   
59945                                            english   

                                                  essay0  \
0      about me:  i would love to think that i was so...   
1      i am a chef: this is what that means. 1. i am ...   
2      i'm not ashamed of much, but writing public te...   
3              i work in a library and go to school. . .   
4      hey how's it going? currently vague on the pro...   
...                                                  ...   
59941  vibrant, expressive, caring optimist. i love b...   
59942  i'm nick. i never know what to write about mys...   
59943  hello! i enjoy traveling, watching movies, and...   
59944  "all i have in this world are my balls and my ...   
59945  is it odd that having a little "enemy" status ...   

                                                  essay1  \
0      currently working as an international agent fo...   
1      dedicating everyday to being an unbelievable b...   
2      i make nerdy software for musicians, artists, ...   
3              reading things written by old dead people   
4                             work work work work + play   
...                                                  ...   
59941  the happiest times have been when life came to...   
59942  currently finishing school for film production...   
59943  i'm a civil engineer, who enjoys helping the c...   
59944  following my dreams... "you got a dream... you...   
59945  i work with elderly people (psychotherapy and ...   

                                                  essay2  \
0      making people laugh. ranting about a good salt...   
1      being silly. having ridiculous amonts of fun w...   
2      improvising in different contexts. alternating...   
3      playing synthesizers and organizing books acco...   
4      creating imagery to look at: http://bagsbrown....   
...                                                  ...   
59941  i make an outstanding osso bucco. i am also ve...   
59942  filmmaking, photography, graphic design, web d...   
59943  - looking at things objectively - getting thin...   
59944                                          listening   
59945  i'm a great bullshitter. i don't know what it ...   

                                                  essay3  \
0      the way i look. i am a six foot half asian, ha...   
1                                                    NaN   
2      my large jaw and large glasses are the physica...   
3                      socially awkward but i do my best   
4                i smile a lot and my inquisitive nature   
...                                                  ...   
59941  i am told that people notice my smile, eyes an...   
59942                                dude, i don't know.   
59943  i'm quiet until i get used to the environment ...   
59944  it used to be the hair until i mowed it off bu...   
59945  either that i am funny/sarcastic, or that i am...   

                                                  essay4  \
0      books: absurdistan, the republic, of mice and ...   
1      i am die hard christopher moore fan. i don't r...   
2      okay this is where the cultural matrix gets so...   
3      bataille, celine, beckett. . . lynch, jarmusch...   
4      music: bands, rappers, musicians at the moment...   
...                                                  ...   
59941  i am an avid movie

In [5]:
df.dtypes

age              int64
status          object
sex             object
orientation     object
body_type       object
diet            object
drinks          object
drugs           object
education       object
ethnicity       object
height         float64
income           int64
job             object
last_online     object
location        object
offspring       object
pets            object
religion        object
sign            object
smokes          object
speaks          object
essay0          object
essay1          object
essay2          object
essay3          object
essay4          object
essay5          object
essay6          object
essay7          object
essay8          object
essay9          object
dtype: object

La codificación de los tipos de las variables es correcta, por lo que no será necesario hacer ninguna transformación de tipos.

 Se tienen variables numéricas y textuales. Sin embargo, observando los contenidos de algunas columnas, es evidente que no son de texto libre, y que se pueden considerar variables categóricas. Vamos a definir los diferentes tipos de características:

In [6]:
numerical_vars = ["age", "height", "income"]
categorical_vars = ["status", "sex", "orientation", "body_type", "diet", "drinks", "drugs", "education",
                    "ethnicity", "job", "location", "offspring", "pets", "religion", "sign", "smokes", "speaks"]
text_vars = [f"essay{i}" for i in range(10)]

Llegados a este punto podemos prescindir de una variable que no será de utilidad a la hora de hacer el clustering. Esta variable es `last_online`, que repesenta la fecha y hora de última conexión. Con esta decisión reducimos el volumen de datos y nos centramos en las columnas numéricas y categóricas que aportan información sobre la personalidad, preferencias o apariencia física de una persona.

In [7]:
del df["last_online"]

### 2.1 Análisis de variables numéricas

#TODO

In [8]:
df.describe()

age        height          income
count  59946.000000  59943.000000    59946.000000
mean      32.340290     68.295281    20033.222534
std        9.452779      3.994803    97346.192104
min       18.000000      1.000000       -1.000000
25%       26.000000     66.000000       -1.000000
50%       30.000000     68.000000       -1.000000
75%       37.000000     71.000000       -1.000000
max      110.000000     95.000000  1000000.000000

En este punto ya podemos observar como hay algunos datos numéricos un tanto extraños:
* La edad máxima es 110, lo cual hace dudar que sea un dato correcto. Posiblemente esta edad corresponda con personas que no quieren compartir su edad verdadera.
* La altura parece estar en pulgadas, *inches*, la media es 68 que se corresponde con 172 cm aprox. En este sentido vemos que la altura mínima es 1 pulgada y que la máxima es de 95 (241 cm). Según la lista de personas más altas del mundo publicada en Wikipedia [1], una altura de entorno a 90 pulgadas(230 cm) ya es algo inaudito.
* Los ingresos llaman la atención en tanto que el ingreso mínimo es -1 y el percentil 75 también es -1. Esto indica que al menos el 75% de los usuarios de la aplicación no han querido decir sus ingresos.

#### 2.2.1 Análisis de la altura

In [9]:
df[df["height"] > 90]["height"]

402      91.0
2569     95.0
3016     95.0
8089     95.0
12181    93.0
17949    95.0
22358    95.0
23760    94.0
25324    95.0
27506    91.0
33640    95.0
33885    95.0
33946    95.0
34607    95.0
34672    94.0
41939    95.0
43452    95.0
44096    95.0
44264    91.0
47621    92.0
48304    95.0
49194    95.0
51714    95.0
53469    95.0
55172    95.0
56931    95.0
57978    95.0
58147    94.0
Name: height, dtype: float64

#### 2.2.2 Análisis de la edad

In [10]:
df[df["age"] > 80]

age     status sex orientation body_type          diet drinks  drugs  \
2512   110     single   f    straight       NaN           NaN    NaN    NaN   
25324  109  available   m    straight  athletic  mostly other    NaN  never   

                        education ethnicity  height  income      job  \
2512                          NaN       NaN    67.0      -1      NaN   
25324  working on masters program       NaN    95.0      -1  student   

                        location        offspring pets  \
2512       daly city, california              NaN  NaN   
25324  san francisco, california  might want kids  NaN   

                                  religion  \
2512                                   NaN   
25324  other and somewhat serious about it   

                                       sign         smokes          speaks  \
2512                                    NaN            NaN         english   
25324  aquarius but it doesn&rsquo;t matter  when drinking  english (okay)   

      essay0 essay1 essay2   essay3 essay4 essay5 essay6 essay7 essay8 essay9  
2512     NaN    NaN    NaN      NaN    NaN    NaN    NaN    NaN    NaN    NaN  
25324    NaN    NaN    NaN  nothing    NaN    NaN    NaN    NaN    NaN    NaN

#### 2.2.3 Análisis de los ingresos

In [11]:
(len(df[df["income"] < 0]) / len(df)) * 100

80.80939512227671

### 2.2 Análisis de las variables categóricas

#### 2.2.1 Análisis del estado marital

El estado marital debería ser una variable categórica simple. Vamos a ver qué posibles valores hay.

In [12]:
df["status"].value_counts(dropna=False)

single            55697
seeing someone     2064
available          1865
married             310
unknown              10
Name: status, dtype: int64

Estos posibles estados parecen ser compatibles entre sí. Asumimos que el estado `unknown` significa que el usuario ha preferido no decirlo, no que sea un tipo diferente. Por eso, podremos codificar esta variable como un `one-hot` haciendo que el valor `unknown` sea un vector completo de ceros. Si hubiera valores `nan`, se considerarían igual que `unknown`.

Parece que no hay valores nulos.

#### 2.2.2 Análisis del sexo

In [13]:
df["sex"].value_counts(dropna=False)

m    35829
f    24117
Name: sex, dtype: int64

En este caso parece una variable binaria clara. Parece que hay una predominancia de hombres en la aplicación.

#### 2.2.3 Análisis de la orientación sexual

In [14]:
df["orientation"].value_counts(dropna=False)

straight    51606
gay          5573
bisexual     2767
Name: orientation, dtype: int64

La variable `orientation` tiene tres posibles valores. Sin embargo, en lugar de clasificar como un vector `one-hot`, creemos que a efectos de agrupar perfiles similares, podemos convertir esta variable en dos columnas binarias nuevas, `likes_men` y `likes_women`. Nos basamos en la idea de que el tipo `straight` no significa lo mismo si la persona en cuestión es hombre que si es mujer. Un clustering funcionará mejor si realizamos esta conversión.

#### 2.2.4 Análisis de la complexión corporal

In [15]:
df["body_type"].value_counts(dropna=False)

average           14652
fit               12711
athletic          11819
NaN                5296
thin               4711
curvy              3924
a little extra     2629
skinny             1777
full figured       1009
overweight          444
jacked              421
used up             355
rather not say      198
Name: body_type, dtype: int64

Hay muchos de estos tipos que podemos agrupar desde el principio. Por ejemplo, `fit` y `athletic`. Como perspectiva alternativa, podemos convertirlo en una o varias escalas ordinales. Por ejemplo, una escala de menor a mayor peso, junto con una escala de menor a mayor proporción de grasa a masa muscular. Si seguimos esta segunda alternativa, el tipo `"rather not say"` tendrá que ser considerado `"average"`.

En cualquiera de los casos, los valores nulos se considerarán iguales a `rather not say`.

#### 2.2.5 Análisis de la dieta

In [16]:
df["diet"].value_counts(dropna=False)

NaN                    24395
mostly anything        16585
anything                6183
strictly anything       5113
mostly vegetarian       3444
mostly other            1007
strictly vegetarian      875
vegetarian               667
strictly other           452
mostly vegan             338
other                    331
strictly vegan           228
vegan                    136
mostly kosher             86
mostly halal              48
strictly kosher           18
strictly halal            18
halal                     11
kosher                    11
Name: diet, dtype: int64

Parece que en este caso tenemos tipos de dieta (`anything, vegan, kosher`) con prefijos que muestran lo estricto que son (`mostly, strictly`). Vamos a ver qué tipos hay:

In [17]:
prefixes = set()
diets = set()
for item in filter(lambda x: not pd.isnull(x), df["diet"].unique()):
    if " " in item:
        prefix, diet = tuple(item.split())
        prefixes.add(prefix)
        diets.add(diet)
    else:
        diets.add(item)
print(prefixes)
print(diets)

{'strictly', 'mostly'}
{'anything', 'kosher', 'halal', 'other', 'vegetarian', 'vegan'}


\# TODO: how to encode this information (for example: vegetarian and vegan should be close)

#### 2.2.6 Análisis de la bebida

In [18]:
df["drinks"].value_counts(dropna=False)

socially       41780
rarely          5957
often           5164
not at all      3267
NaN             2985
very often       471
desperately      322
Name: drinks, dtype: int64

Esto parece un ejemplo claro de variable ordinal. Hay demasiados valores `NaN` como para eliminarlos, pero podremos utilizar un `imputer` para completar los datos que nos faltan.

#### 2.2.7 Análisis de la toma de drogas

In [19]:
df["drugs"].value_counts(dropna=False)

never        37724
NaN          14080
sometimes     7732
often          410
Name: drugs, dtype: int64

De nuevo, esto parece ser un ejemplo de variable ordinal. Sin embargo, aquí tenemos una hipótesis que comprobaremos más adelante sobre los valores `NaN`: que los valores `NaN` deberían estar en la escala ordinal entre `never` y `sometimes`, porque son personas que han tomado drogas esporádicamente en el pasado, pero que no quieren decir `sometimes`. 

\# TODO: see if we can prove this

#### 2.2.8 Análisis de la educación recibida

In [20]:
df["education"].value_counts(dropna=False)

graduated from college/university    23959
graduated from masters program        8961
NaN                                   6628
working on college/university         5712
working on masters program            1683
graduated from two-year college       1531
graduated from high school            1428
graduated from ph.d program           1272
graduated from law school             1122
working on two-year college           1074
dropped out of college/university      995
working on ph.d program                983
college/university                     801
graduated from space camp              657
dropped out of space camp              523
graduated from med school              446
working on space camp                  445
working on law school                  269
two-year college                       222
working on med school                  212
dropped out of two-year college        191
dropped out of masters program         140
masters program                        136
dropped out

Esto parece otro ejemplo de variables con prefijos. Observando las combinaciones únicas que tenemos, podemos extraer los prefijos:

In [21]:
prefixes = ["graduated from", "working on", "dropped out of"]
educations = set()
for item in filter(lambda x: not pd.isnull(x), df["education"].unique()):
    for prefix in prefixes:
        if prefix in item:
            educations.add(item.replace(prefix, "").strip())
print(educations)

{'space camp', 'college/university', 'high school', 'two-year college', 'med school', 'law school', 'masters program', 'ph.d program'}


### 2.3 Análisis de las variables de texto libre

## 3. Preprocesamiento

## 4. Clustering

## 5. Conclusiones

## 6. Referencias

[1] https://en.wikipedia.org/wiki/List_of_tallest_people

---
Creado por **Diego Trujillo** (diego.trujillo.jimenez@alumnos.upm.es) y **Ángel Acevedo** (angel.acevedo.sanchez@alumnos.upm.es) 

<br></br>
[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)